### ISSUE TITLE
searching through multiple indices

### ISSUE THREAD

**Issue by psaiz** [sscarduzio/elasticsearch-readonlyrest-plugin#178]
Hi, 

I'm not sure if this is a bug, or a design choice. Imagine that I had two rules, like  :
```
- name: "Allow index1"
  type: allow
  indices:
    - my_test1
- name: "Allow index2"
  type: allow
  indices:
    - my_test_2
```
If I do a search on my_test1, it works nicely. If I do a search on my_test2, it also works nicely. The issue is if I do a query on my_test*. I would expect this to return the documents from both indices, but it only returns the documents from my_test1, since that rule matches first. Is that intended?

Cheers,
pablo

**Comment by sscarduzio** [comment]
Very interesting one. A bit counter intuitive, but the system behaves as specs: exit at first match.
Although this case is trivially resolved with

```yml
- name: "Allow index1-2"
  type: allow
  indices:
    - my_test1
    - my_test2
```
Another case could be not as easy to reconcile:
```yml
- name: "Allow index1"
  type: allow
  hosts: 1.1.1.1
  indices:
    - my_test1

- name: "Allow index2"
  type: allow
  api_keys: 123123
  indices:
    - my_test_2
```
With the above, when a request comes from 1.1.1.1 and carries the right api key, one would expect  documents from both indices are returned.

In principle, we could change this: instead of stopping evaluation at the first match, we could evaluate all remaining rules IFF THE FIRST MATCHED BLOCK IS "ALLOW".  Therefore allowing all side effects (granted indices and set cookies) to pile up and be committed to the REST request object.

I don't see problems for indices, but I think it won't work for cookies.

**Comment by sscarduzio** [comment]
I'm actually working on this right now, kinda big change, but may also fix also the groups issue  #165

**Comment by psaiz** [comment]
Hi, 
Thanks for looking into this. What would happen if the first rule is forbid? I mean, imagine something like
```
- name: "Forbid private index"
  indices: 
   - my_private
  type: forbid
- name: "Allow public"
  indices: 
    - my_public
  type: allow
```

If I query my_*, would I get a forbid, or would I get the documents from my_public?

Cheers, 
pablo

**Comment by sscarduzio** [comment]
I've slept over this one. So here it boils down to two alternative algorithms:

### Chain model (current)
Serial evaluation of blocks, exiting on the first matched block and applying the block's policy (forbid or allow). Apply only the side effects (indices list expansion/modification, cookie set) of the matching block.

### Survivors model
Parallel evaluation of all the blocks. Apply the side effects (indices list expansion/modification) of the "survivor" (matching) blocks in the order as they were declared. If a survivor block has forbid policy, the indices therein declared will be removed from the indices list  rather than retained.

I think we should **keep the "chain" model**. It saves the day in terms of ease of reasoning.
Of course: order of the blocks will continue to matter, at times it will make necessary to duplicate some rules for same users (when you want to express OR logic), but it does a great job at protecting us from very costly mistakes.

*Human error is the top cause in security flaws, simplicity of configuration, familiarity and minimal ambiguity are paramount. I don't care of a 10% more powerful DSL.*

Basically all firewalls and LDAP work like this.

There's always the rule of thumb of sorting the rules from least to most specific, works well, it will be all right.

---

Finally, to answer your question:

### Current chain logic with side effects
* A request referred to my_* will be expanded to all indices in the cluster that start with "my_". In this case (I suppose): 
`req(my_*) => req(my_private, my_public).`

* The first block will match because my_private is in the indices list. Also, only the matched indices will be retained:
` req(my_private,my_public) => req(my_private)`

* The ACL will terminate because we have a match: request is forbidden.

**Comment by sscarduzio** [comment]
I've been thinking that we could add a configuration key that specifies an alternative ACL logic:

```yml
readonlyrest:
    enable: true
    logic: parallel #defaults to serial

    access_control_rules:
    - name: "...bla"
      type: allow
      hosts: [127.0.0.1]
```
